packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/api-client/iroha-api-client.ts

```diff
@@ -0,0 +1,261 @@
+    this.log.debug(monitorOptions);
+    socket.on(WatchBlocksV1.Next, (data: IrohaBlockProgress) => {
+      subject.next(data);
+    });
+
+    socket.on("connect", () => {
+      this.log.debug("connected OK...");
+      socket.emit(WatchBlocksV1.Subscribe, monitorOptions);
+    });
+
+    socket.connect();
+
+    socket.on("connect_error", (err: Error) => {
+      this.log.error("Error (connect_error): ", err);
+      socket.disconnect();
+      subject.error(err);
+    });
+
+    socket.on("connect_timeout", (err: Record<string, unknown>) => {
+      this.log.error("Error (connect_timeout): ", err);
+      socket.disconnect();
+      subject.error(err);
+    });
+
+    socket.on("error", (err: unknown) => {
+      this.log.error("Error: ", err);
+      socket.disconnect();
+      subject.error(err);
+    });
+
+    return subject.pipe(
+      finalize(() => {
+        this.log.info("FINALIZE - unsubscribing from the stream...");
+        socket.emit(WatchBlocksV1.Unsubscribe);
+        socket.disconnect();
+      }),
+      share(),
+    );
+  }
+
+  /**
+   * Immediately sends request to the validator, doesn't report any error or responses.
+   * @param args - arguments.
+   * @param method - function / method to be executed by validator.
+   * @param baseConfig - baseConfig needed to properly connect to ledger
+
+   */
+  public sendAsyncRequest(
+    method: Record<string, unknown>,
+    args: any,
+    baseConfig?: IrohaBaseConfig,
+  ): void {
+    this.log.debug(`inside: sendAsyncRequest()`);
+    this.log.debug(`baseConfig=${baseConfig}`);
+    this.log.debug(`methodName=${method.methodName}`);
+    this.log.debug(`args=${args}`);
+
+    if (!baseConfig) {
+      throw new RuntimeError("baseConfig object must exist and not be empty");
+    }
+
+    Checks.truthy(baseConfig.privKey, "privKey in baseConfig");
+    Checks.truthy(
+      baseConfig.creatorAccountId,
+      "creatorAccountId in baseConfig",
+    );
+    Checks.truthy(baseConfig.irohaHost, "irohaHost in baseConfig");
+    Checks.truthy(baseConfig.irohaPort, "irohaPort in baseConfig");
+    Checks.truthy(baseConfig.quorum, "quorum in baseConfig");
+    Checks.truthy(baseConfig.timeoutLimit, "timeoutLimit in baseConfig");
+    Checks.nonBlankString(method.methodName, "methodName");
+
+    const socket: Socket = io(this.wsApiHost, { path: this.wsApiPath });
+    const asyncRequestData = {
+      baseConfig: baseConfig,
+      methodName: method.methodName,
+      args: args,
+    };
+
+    this.log.debug("requestData:", asyncRequestData);
+
+    try {
+      socket.emit(IrohaSocketIOTransactV1.SendAsyncRequest, asyncRequestData);
+
+      // Connector should disconnect us after receiving this request.
+      // If he doesn't, disconnect after specified amount of time.
+      setTimeout(() => {
+        if (socket.connected) {
+          socket.disconnect();
+        }
+      }, baseConfig.timeoutLimit ?? 10 * 1000);
```

Comment:
I think this wrongly connects `sendAsyncRequest` from other ledger in discounted-asset-trade sample with this ApiClient and reports error.

Reply:
> outSH4 days ago
> I think this wrongly connects sendAsyncRequest from other ledger in discounted-asset-trade sample with this ApiClient and reports error.

@outSH From what I gather, the timeout is a request parameter that the caller can specify to be anything when coming in through `PluginLedgerConnectorIroha#transact()` (which ultimately ends up running this code if I read it right).
And it's saying that I could send requests with the timeout being set to a decade and then bring down the servers by sending a lot of those requests that allocate memory (technically) forever through the setTimeout.
A few things that you could do IMO:
1. Specify in the openapi.json that there is a maximum value for the timeout, something like 3 minutes or maybe 10 (hard to tell in the end because you don't want to run into issues with it being too short either...)
2. Validate here in the code that the timeout is not longer than what's allowed so that if someone tries to set it to a decade, they won't succeed.
3. Add a description to the timeout field that explains that it is to be specified in milliseconds and/or rename the field to `timeoutLimitMs`.
4. If after 1) and 2) and 3) CodeQL still complains then I think we are justified to mark it as "resolved" manually, but right now it looks like it is a legit concern, assuming that `baseConfig.timeoutLimit` is indeed something the requestors can specify themselves to be anything.

Reply:
@petermetz Not really, this is an `ApiClient` class used by the client and it sets the timeout on the client side. The only leak is socketio connection, but this can be caused as easily without timeout at all (sort of affects all the connectors at the moment).

The problem that CodeQL catches here is example app, which blindly passes the args to `sendSyncMethod`, so a person who uses BLP can cause BLP to leak timeouts - but it doesn't, since it uses different ledger apiClient, not this one, and in general should never be done in an actual app for various other security reasons.

Nevertheless I will change it anyway to match the other ApiClients, it seems to be a better practice to set this timeout in ApiClient constructor to better isolate it from any possible BLP-user input.